#26520 Add siglip attention module and support to TTT (gemma3 vision)

handrewsTT opened the pull request:
### Ticket
#26520

### Problem description
We need to modify the TTT llama vision attention to support siglip attention.

### What's changed
Added bias to qkv and output linear projections and added the ability to pass in sdpa config in the configuration

Also added gemma3 tests to t3k unit test and single card frequent-fast-dispatch pipelines.

### Checklist
[APC](https://github.com/tenstorrent/tt-metal/actions/runs/17042018146)
[T3K unit tests](https://github.com/tenstorrent/tt-metal/actions/runs/16990677853)
[Single card (demo + frequent)](https://github.com/tenstorrent/tt-metal/actions/runs/16990680216) (frequent rerun: https://github.com/tenstorrent/tt-metal/actions/runs/17044151596)

ppetrovicTT committed:
Global and local embeddings

handrewsTT committed:
rebase and resolve conflicts

handrewsTT committed:
add linear scaling support to rope

handrewsTT committed:
update to new transformer forward API

handrewsTT committed:
fix precommit

handrewsTT committed:
remove unwanted test

handrewsTT committed:
fix attention test

handrewsTT committed:
fix llama vision api

handrewsTT committed:
passing siglip sdpa

handrewsTT committed:
add siglip tests and license headers

ppetrovicTT reviewed (changes requested):
For tests, I would prefer if we took the route from tt_transformers - comparing with the HF transformers implementation directly. This way we don't need the reference code at all (@yieldthought told me this approach is deprecated).

We run TT verison, than imported HF version and compare PCC.

handrewsTT commented:
> For tests, I would prefer if we took the route from tt_transformers - comparing with the HF transformers implementation directly. This way we don't need the reference code at all (@yieldthought told me this approach is deprecated).
> 
> We run TT verison, than imported HF version and compare PCC.

The test does actually compare directly to the HF transformers implementation, the "reference" code is more there as a sanity check. I have added a copy of model.py which contains the HF model but this allows us to decorate the `forward` method and capture intermediate outputs easily which is really helpful for development speed. Once we've got the full model working e2e we can definitely remove this.

ppetrovicTT commented on models/demos/siglip/tests/test_attention.py:
```diff
@@ -0,0 +1,62 @@
+@pytest.mark.parametrize("attention_func", [siglip_attention, siglip_attention_ttnn])
+def test_attention_rand(attention_func):
+    with open(
+        "/proj_sw/user_dev/hf_data/hub/models--google--gemma-3-4b-it/snapshots/093f9f388b31de276ce2de164bdc2081324b9767/config.json",
```
we can't use /proj_sw/ in CI i think, but should stick to /mnt/MLPerf or whatever is there for CIv2.

however, i think we should use the transformers library to get this config, as it might contain some hardcoded values that don't exist in the config.json. e.g. `Gemma3Config.vision_config`

handrewsTT replied:
I am planning to update /proj_sw once @mtairum has copied the model to `/mnt/MLPerf`

handrewsTT replied:
What do you mean get the config from the transformers library? I thought we always provided model configurations from the config.json in models on the huggingface hub that we store locally?

ppetrovicTT replied:
In model_config.py we're actually using the Config from transformers to extract the info (and it is loading the config.json).

Since here you're trying to read the config directly, without loading the model, I'd suggest that you use AutoConfig.from_pretrained and get the actual config from Transformers.

handrewsTT committed:
add siglip unit tests to CI

ppetrovicTT commented on models/demos/siglip/tests/test_attention.py:
```diff
@@ -17,12 +17,10 @@
-        config = SiglipVisionConfig(**json.load(fp)["vision_config"])
+def test_attention(attention_func):
+    config = AutoConfig.from_pretrained(os.getenv("HF_MODEL"))
+    if hasattr(config, "vision_config"):
```
i think we need to assert this

ppetrovicTT commented on tests/scripts/run_python_model_tests.sh:
```diff
@@ -81,6 +81,9 @@ run_python_model_tests_wormhole_b0() {
     mistral_weights=mistralai/Mistral-7B-Instruct-v0.3
     HF_MODEL=$mistral_weights pytest -n auto models/tt_transformers/tests/test_model.py -k "quick" ; fail+=$?
+
+    # Siglip
```
Siglip (Gemma3 Vision)

handrewsTT committed:
update hf path

handrewsTT committed:
update hf path

handrewsTT committed:
install transformers dependency for tests

handrewsTT committed:
create separate development tests job

handrewsTT committed:
change import

handrewsTT committed:
rebase and resolve conflicts

handrewsTT committed:
update ccls and device logic for attention tests

handrewsTT committed:
fix for new ccls

handrewsTT committed:
remove accidental files

handrewsTT committed:
precommit

handrewsTT committed:
fix tests

handrewsTT committed:
remove unneeded code

handrewsTT committed:
make tests compatible with newer transformers version

handrewsTT committed:
remove unneeded requirements

handrewsTT committed:
remove requirements install from tests

handrewsTT committed:
explicit gemma path

handrewsTT committed:
use central mesh device creation

handrewsTT committed:
gemma CI test changes

### Ticket
Link to Github Issue

### Problem description
Provide context for the problem.

### What's changed
Describe the approach used to solve the problem.
Summarize the changes made and its impact.

### Checklist
- [ ] [All post
commit](https://github.com/tenstorrent/tt-metal/actions/workflows/all-post-commit-workflows.yaml)
CI passes
- [ ] [Blackhole Post
commit](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml)
CI with demo tests passes (if applicable)
- [ ] [Model
regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-models.yaml)
CI passes (if applicable)
- [ ] [Device performance
regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-device-models.yaml)
CI passes (if applicable)
- [ ] (For models and ops writers) [Single-card demo
tests](https://github.com/tenstorrent/tt-metal/actions/workflows/single-card-demo-tests.yaml)
CI passes (if applicable) See [recommended dev
flow](https://github.com/tenstorrent/tt-metal/blob/main/models/docs/MODEL_ADD.md#a-recommended-dev-flow-on-github-for-adding-new-models).
- [ ] [Galaxy
quick](https://github.com/tenstorrent/tt-metal/actions/workflows/tg-quick-trigger.yaml)
CI passes (if applicable)
- [ ] [TG demo tests, for
Llama](https://github.com/tenstorrent/tt-metal/actions/workflows/tg-demo-tests.yaml)
CI passes, if applicable, because of current Llama work
- [ ] (For runtime and ops writers) [T3000 unit
tests](https://github.com/tenstorrent/tt-metal/actions/workflows/t3000-unit-tests.yaml)
CI passes (if applicable, since this is run on push to main)
- [ ] (For models and ops writers) [T3000 demo
tests](https://github.com/tenstorrent/tt-metal/actions/workflows/t3000-demo-tests.yaml)
CI passes (if applicable, since this is required for release)
- [ ] (For models and ops writers) [Single-card demo
tests](https://github.com/tenstorrent/tt-metal/actions/workflows/single-card-demo-tests.yaml)
CI passes (if applicable, since this is required for release)
- [ ] New/Existing tests provide coverage for changes

---------

Co-authored-by: Pavle Petrovic <ppetrovic@tenstorrent.com>

handrewsTT committed:
add single card unit tests for gemma

handrewsTT committed:
TEMP: update single card fast dispatch workflow

handrewsTT committed:
remove commented out frequent models tests

ppetrovicTT commented on models/demos/gemma3/tests/test_ci_dispatch.py:
```diff
@@ -0,0 +1,42 @@
+        "/mnt/MLPerf/huggingface/hub/models--google--gemma-3-4b-it/snapshots/093f9f388b31de276ce2de164bdc2081324b9767",
+    ],
+    ids=[
+        "gemma3-4b-it",
```
please mind the dashes:
gemma-3-4b-it

ppetrovicTT commented on models/demos/gemma3/tests/test_ci_dispatch.py:
```diff
@@ -0,0 +1,42 @@
+@pytest.mark.parametrize(
+    "model_weights",
+    [
+        "/mnt/MLPerf/huggingface/hub/models--google--gemma-3-4b-it/snapshots/093f9f388b31de276ce2de164bdc2081324b9767",
```
is there a way that we can use "google/gemma-3-4b-it" here and set HF_HOME to the proper location? i find it cleaner to rely on HF ecosystem.

ppetrovicTT commented on models/demos/gemma3/tests/test_ci_dispatch.py:
```diff
@@ -0,0 +1,42 @@
+def test_ci_dispatch(model_weights):
+    logger.info(f"Running fast dispatch tests for {model_weights}")
+    if os.getenv("HF_MODEL"):
+        del os.environ["HF_MODEL"]
```
i think we don't have to delete these if we're overriding them below anyway

handrewsTT committed:
fast dispatch gemma test changes

handrewsTT committed:
conform to meta attention weights format

handrewsTT committed:
fix reference attention test

tt-rkim commented on .github/workflows/single-card-demo-tests-impl.yaml:
```diff
@@ -221,6 +221,8 @@ jobs:
             -e TT_METAL_HOME=${{ github.workspace }}
             -e LD_LIBRARY_PATH=${{ github.workspace }}/build/lib
             -e ARCH_NAME=${{ inputs.arch }}
+            -e HF_HUB_OFFLINE=1
```
Is this something we want to start doing for all pipelines? Is this what @gwangTT has beend goin?

tt-rkim commented on .github/workflows/t3000-fast-tests-impl.yaml:
```diff
@@ -31,6 +31,8 @@ jobs:
         LD_LIBRARY_PATH: /work/build/lib
         ARCH_NAME: ${{ matrix.test-group.arch }}
+        HF_HUB_OFFLINE: 1
+        HF_HOME: /mnt/MLPerf/huggingface
```
Note that this will be diff in CIv2

handrewsTT replied:
Will these workflows be running on Clv2? Sorry I'm not totally up to speed on the CI infra

tt-rkim commented on .github/workflows/t3000-unit-tests-impl.yaml:
```diff
@@ -41,6 +41,7 @@ jobs:
           { name: "t3k unet shallow tests", arch: wormhole_b0, cmd: run_t3000_unet_shallow_tests, timeout: 30, owner_id: U06ECNVR0EN}, #Evan Smal
           { name: "t3k mistral tests", arch: wormhole_b0, cmd: run_t3000_mistral_tests, timeout: 30, owner_id: U0896VBAKFC}, # Pratikkumar Prajapati
           { name: "t3k qwen25_vl tests", arch: wormhole_b0, cmd: run_t3000_qwen25_vl_unit_tests, timeout: 40, owner_id: U07RY6B5FLJ}, #Gongyu Wang
+          { name: "t3k gemma3-small tests", arch: wormhole_b0, cmd: run_t3000_gemma3-small_tests, timeout: 30, owner_id: U08E1JCDVNX},  #Pavle Petrovic
```
Pavle owns these tests?

handrewsTT replied:
These tests don't need weights but they need the model config which is stored in the checkpoint. These tests take ~100s to run on my machine

handrewsTT replied:
Pavle actually added these to t3k unit tests and he is leading the gemma text model development which these tests will largely capture when that work merges I think. So it makes sense for either him or me to own these tests

tt-rkim commented on models/demos/gemma3/tests/test_ci_dispatch.py:
```diff
@@ -0,0 +1,41 @@
+)
+def test_ci_dispatch(model_weights):
+    logger.info(f"Running fast dispatch tests for {model_weights}")
+    os.environ["HF_HOME"] = "/mnt/MLPerf/huggingface"
```
Not sure about hard setting these in here if you're gonna set them in the workflow anyway. We should be consistent.

Why don't you start integrating this stuff into `model_location_generator` in `conftest.py`?

handrewsTT commented on tests/scripts/t3000/run_t3000_unit_tests.sh:
```diff
@@ -163,6 +163,25 @@ run_t3000_falcon40b_tests() {
   fi
 }
 
+run_t3000_gemma3-small_tests() {
+  # Record the start time
+  fail=0
+  start_time=$(date +%s)
+
+  echo "LOG_METAL: Running run_t3000_gemma3-small_tests"
+
+  HF_MODEL="google/gemma-3-4b-it" TT_CACHE_PATH="$HF_HOME/tt_cache/gemma-3-4b-it" pytest -n auto models/demos/siglip/tests ; fail+=$?
+  echo "LOG_METAL: $HF_MODEL tests completed"
+
+  # Record the end time
+  end_time=$(date +%s)
+  duration=$((end_time - start_time))
+  echo "LOG_METAL: run_t3000_gemma3-small_tests $duration seconds to complete"
+  if [[ $fail -ne 0 ]]; then
+    exit 1
+  fi
+}
+
```
Gotcha, I'll make the change.

Siglip is a model in it's own rights, with vision and text. We are currently just standing up the vision part as gemma3 uses it for vision encodings more multimodal models. The tests I'm trying to include here a unit tests that check PCC for the individual components we need and the idea is that when stuff breaks while we're in the process of bringing up the model we can catch it early.

handrewsTT committed:
review comments

handrewsTT committed:
adding missing dispatch env vars
